server/main_server/src/main.cpp

```diff
@@ -1450,6 +1488,273 @@ Environment Variables:
+
+            // Allow the admin to limit the number of atime updates in case the queue size is very large.
+            const auto update_count =
+                std::min(irods::access_time_queue::number_of_queued_updates(), g_atime_batch_size);
+            log_server::debug(
+                "{}: Number of access time updates before deduplication is [{}].", __func__, update_count);
+
+            for (std::size_t i = 0; i < update_count; ++i) {
+                if (!irods::access_time_queue::try_dequeue(data)) {
+                    break;
+                }
+
+                // clang-format off
+                updates.push_back(nlohmann::json{
+                    {"data_id", data.data_id},
+                    {"replica_number", data.replica_number},
+                    {"atime", data.last_accessed}
+                });
+                // clang-format on
+            }
+
+            if (updates.empty()) {
+                return;
+            }
+
+            // Group elements by data id and replica number in descending order.
+            std::sort(std::begin(updates), std::end(updates), [](const auto& _lhs, const auto& _rhs) {
+                const auto& lhs_data_id = _lhs.at("data_id");
+                const auto& rhs_data_id = _rhs.at("data_id");
+                if (lhs_data_id > rhs_data_id) {
+                    return true;
+                }
+
+                if (lhs_data_id == rhs_data_id) {
+                    const auto& lhs_replica_number = _lhs.at("replica_number");
+                    const auto& rhs_replica_number = _rhs.at("replica_number");
+                    if (lhs_replica_number > rhs_replica_number) {
+                        return true;
+                    }
+
+                    if (lhs_replica_number == rhs_replica_number) {
+                        return _lhs.at("atime") > _rhs.at("atime");
+                    }
+                }
+
+                return false;
+            });
+
+            // Reduce updates such that each replica is updated once. On completion, the JSON
+            // is guaranteed to contain only the most recent update for each replica.
+            auto last = std::unique(std::begin(updates), std::end(updates), [](const auto& _lhs, const auto& _rhs) {
+                return _lhs.at("data_id") == _rhs.at("data_id") &&
+                       _lhs.at("replica_number") == _rhs.at("replica_number");
+            });
+            updates.erase(last, std::end(updates));
+            log_server::debug(
+                "{}: Deduplication complete. Access time updates reduced to [{}].", __func__, updates.size());
+
+            // Apply updates.
+            const auto json_input = nlohmann::json{{"access_time_updates", updates}}.dump();
+            char* ignored{};
+            const auto ec = rc_update_replica_access_time(static_cast<RcComm*>(conn), json_input.c_str(), &ignored);
+            log_server::debug("{}: rc_update_replica_access_time returned [{}].", __func__, ec);
```

Comment:
Do we want to log a warning or info message if `ec < 0`?

Reply:
On second thought, debug may be the most appropriate. At most, info.

Reply:
Hmm, debug feels correct to me, but I can see toggling the error category based on `ec < 0`. I'd imagine that if there was a database issue, the admin and users would see it in other places too.

Thoughts?

Reply:
I think debug is fine. I was initially thinking that it would be good for finding strange issues, but it may be indicative of a general database issue as you said, or it could be indicative of someone messing with the data object (open-for-read/close, then quickly unlink the replica before the atime can be updated). Might just get noisy.

Reply:
Yep. Leaving as is.